### ISSUE TITLE
Elsa 2.0: Consideration to use AutoMapper.Extensions.Microsoft.DependencyInjection, instead of own implementation

### ISSUE THREAD

**Issue by Rafael94** [elsa-workflows/elsa-core#535]
I have problems with my AutoMapper profiles. As soon as I use `AddAutoMapper` from Elsa Core, my profiles are not found. So I replaced it with an empty function on my end and call the AutoMapper functions in my main application.

I guess it is due to the AutoMapper implementation of Elsa Core. I use `AutoMapper.Extensions.Microsoft.DependencyInjection`.

I could replace the implementation of Elsa with `AutoMapper.Extensions.Microsoft.DependencyInjection`. But I wanted to inquire first if it is desired or if someone has another idea.

**Comment by sfmskywalker** [comment]
The easiest way is to use the extension method provided by Elsa (`AddAutoMapperProfile`) to add your own profiles. Elsa already registers AutoMapper, so you will want to add your profiles to that instance.

Alternatively, you can configure Elsa to use a different AutoMapper instance, but then you need to make sure to also register Elsa’s profiles.

Perhaps it would be better if we simply make AutoMapper a private dependency and wrap it in a custom mapper class that is them registered with DI. This way it won’t interfere with custom applications.

**Comment by Rafael94** [comment]
That would also be a possibility. Do you mean a separate class that implements `IMapper`?

But I guess `AutoMapper.Extensions.Microsoft.DependencyInjection` would already fix the problem.

**Comment by sfmskywalker** [comment]
It would be a separate abstraction altogether as to not conflict with the instance that the extension you mentioned would register.

Using said extension would be preferred Le, but there are two limitations with it:

1. It does not allow you to register singleton mappers
2. There is no way to register additional profiles from other class libraries. For example, how would the YesSQL persistence provider add its mapping profiles?

It’s been a while since I looked into this, so perhaps things have changed a bit. If you do see a way for Elsa and it’s satellite class libraries to register mapping profiles without adding multiple AutoMapper instances, then I’d be all for it.

**Comment by Rafael94** [comment]
What do you mean by "It does not allow you to register singleton mappers"? Are there multiple mappers? You can register the default mapper as a singleton using a parameter.

Additional profiles can be registered with `services.Configure<MapperConfigurationExpression>(options => options.AddMaps(assembliesToScanArray));`. I could write an extension method for this to make it more understandable to use.

AddAutoMapper can be called multiple times without any problems.  
```
// Just return if we've already added AutoMapper to avoid double-registration
if (services.Any(sd => sd.ServiceType == typeof(IMapper)))
  return services;
```

**Comment by sfmskywalker** [comment]
I meant that `AutoMapper.Extensions.Microsoft.DependencyInjection` doesn't (or at least didn't) support registering IMapper as a singleton (see [this thread](https://github.com/AutoMapper/AutoMapper.Extensions.Microsoft.DependencyInjection/issues/81#issuecomment-486224508))

But that's not a big issue per se. It's arguably a small optimization to not have to instantiate a new mapper per scope.

But I am excited to hear about the ability to call AddAutoMapper multiple times and being able to append more profiles. That would allow anyone to use both `Elsa` and `AutoMapper.Extensions.Microsoft.DependencyInjection`, where Elsa can register profiles against the same IMapper instances as the host application itself.

**Comment by Rafael94** [comment]
> I meant that `AutoMapper.Extensions.Microsoft.DependencyInjection` doesn't (or at least didn't) support registering IMapper as a singleton (see [this thread](https://github.com/AutoMapper/AutoMapper.Extensions.Microsoft.DependencyInjection/issues/81#issuecomment-486224508))
> 
> But that's not a big issue per se. It's arguably a small optimization to not have to instantiate a new mapper per scope.
> 
> But I am excited to hear about the ability to call AddAutoMapper multiple times and being able to append more profiles. That would allow anyone to use both `Elsa` and `AutoMapper.Extensions.Microsoft.DependencyInjection`, where Elsa can register profiles against the same IMapper instances as the host application itself.

If I understand the code snippet correctly, they changed it.

```
services.Add(new ServiceDescriptor(typeof(IMapper),
                sp => new Mapper(sp.GetRequiredService<IConfigurationProvider>(), sp.GetService), serviceLifetime));
```

`serviceLifetime` is a parameter that can be specified.

**Comment by sfmskywalker** [comment]
Great. Thanks for the PR!

**Comment by sfmskywalker** [comment]
Fixed in 5b84a2e7a2e0cce33501fcd02f36f2411be68619
